File: leetcode-go/Q001-Q050/Q033_Search_in_Rotated_Sorted_Array_B/Q033_Search_in_Rotated_Sorted_Array_B.go
package Q033_Search_in_Rotated_Sorted_Array_B

func search(nums []int, target int) int {
	l, h := 0, len(nums)-1
	for l <= h {
		m := (l + h) / 2

		if nums[m] == target {
			return m
		}
		if nums[l] < nums[m] {
			if target >= nums[l] && target < nums[m] {
				h = m //- 1
			} else {
				l = m + 1
			}
		} else if nums[l] > nums[m] {
			if target > nums[m] && target <= nums[h] {
				l = m + 1
			} else {
				h = m //- 1
			}
		} else {
			l++
			//h--
		}
	}
	return -1
}

File: leetcode-java/src/archive/linkedin/Q033_Search_in_Rotated_Sorted_Array_B.java
package archive.linkedin;

public class Q033_Search_in_Rotated_Sorted_Array_B {

    public int search(int[] nums, int target) {
        int l = 0, r = nums.length-1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (nums[m] == target) {
                return m;
            }
            if (nums[l] <= nums[m]) { // first half is in order
                if (nums[l] <= target && target < nums[m]) {
                    r = m-1;
                } else {
                    l = m+1;
                }
            } else {
                if (nums[m] < target && target <= nums[r]) {
                    l = m+1;
                } else {
                    r = m-1;
                }
            }
        }
        return -1;
    }

}
